### PR TITLE
perf: faster multimodal assistant-only masking (vectorized scanner)

### PR DESCRIPTION
**Pull request by thad0ctor**
# Description

Follow-up to [#3625](https://github.com/axolotl-ai-cloud/axolotl/pull/3625) reducing the cost of multimodal assistant-only loss masking.

- Vectorized variant of the role-boundary scanner, gated on `torch.get_num_threads() == 1` (DataLoader-worker default). Byte-identical output to the reference scanner.
- Fused `process_labels`: role/pad/media masks combined into one boolean keep tensor with a single `labels[~keep] = -100` write.
- One-shot warning when `MultiModalChatDataCollator` is built with `dataloader_num_workers in (None, 0)` and masking is enabled.

## Motivation and Context

@NanoCode012 requested a performance review of #3625. Real-tokenizer bench under DataLoader-worker conditions vs updated vectorize approach:

| Tokenizer | bs=8 seq~3k | bs=8 seq~6k | Speedup |
|---|---|---|---|
| Gemma 3 | ref 4.41 → vec 2.89 ms | ref 8.75 → vec 5.87 ms | 1.49–1.53× |
| Gemma 4 | ref 4.44 → vec 3.35 ms | ref 8.73 → vec 6.53 ms | 1.33–1.36× |
| Qwen 2  | ref 4.46 → vec 3.29 ms | ref 8.77 → vec 6.45 ms | 1.35–1.36× |
| Llama 3 | ref 4.45 → vec 4.42 ms | ref 8.76 → vec 8.68 ms | tie |

Native `return_assistant_tokens_mask=True` was investigated and rejected: silent failure on right-padded multi-row batches in `ProcessorMixin.apply_chat_template` (bisect over non-monotonic offset_mapping), and ~10× slower end-to-end than the scanner.

## How has this been tested?

`pytest tests/test_processing_strategies.py tests/test_vectorized_scanner.py tests/test_process_labels_fusion.py tests/test_mm_collator_warnings.py` — 572 passing.

- `test_vectorized_scanner.py` — 2,000-config differential fuzz + edge cases (Pixtral `[/INST]` rewind, longest-prefix tie, `train_on_eos` modes, `include_end` leak gate).
- `test_process_labels_fusion.py` — 450 fuzz cases across all 9 `process_labels`-overriding strategies asserting `torch.equal` vs pre-refactor.
- successfully started Qwen3-VL masked multimodal training on forked venv 

## AI Usage Disclaimer

Yes. Claude Code (Opus 4.7) assisted throughout. 

## Types of changes

- [x] Performance / refactor
- [x] Tests

## Summary by CodeRabbit

* **New Features**
  * Added one-time warning notification for multimodal training when dataloader worker configuration is unset or zero.

* **Improvements**
  * Enhanced label processing with optimized masking strategies and vectorized operations for improved efficiency across model variants.

* **Tests**
  * Added comprehensive test suites for configuration warnings, label processing parity validation across multiple model strategies, and vectorized masking operations with edge-case and differential fuzzing coverage.

[Review Change Stack](https://app.coderabbit.ai/change-stack/axolotl-ai-cloud/axolotl/pull/3672?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: tests/test_process_labels_fusion.py [hunk 1]
```diff
@@ -0,0 +1,539 @@
...
+        (_make_llama32v, legacy_base_process_labels, [], "llama32v"),
+        (_make_llama4, legacy_base_process_labels, [], "llama4"),
+        (_make_pixtral, legacy_base_process_labels, [], "pixtral"),
+        (_make_mistral_v7, legacy_base_process_labels, [], "mistral_v7"),
+    ],
+)
+@pytest.mark.parametrize("seed", list(range(50)))
+def test_process_labels_parity_fuzz(strategy_factory, legacy_fn, extra_ids, name, seed):
+    """50 random batches per strategy: legacy vs fused must produce byte-identical labels."""
+    rng = random.Random(seed * 7919 + hash(name) % 1000)
```
[the remark this answers is not in the document]
Comment (thad0ctor) [reply tests/test_process_labels_fusion.py]: Addressed in follow on commit

File: tests/test_mm_collator_warnings.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+def test_warn_num_workers_zero_for_mm_is_one_shot(_reset_mm_warn_state, caplog):
+    """Second invocation in the same process should NOT re-emit the warning."""
+    cfg = _make_cfg(processor_type="qwen2_vl", dataloader_num_workers=0)
+    first_msgs = _capture_warning(caplog, cfg)
+    assert any("dataloader_num_workers=0" in m for m in first_msgs)
+
+    caplog.clear()
+    second_msgs = _capture_warning(caplog, cfg)
+    assert not any("dataloader_num_workers=0" in m for m in second_msgs)
```
[the remark this answers is not in the document]
Comment (thad0ctor) [reply tests/test_mm_collator_warnings.py:82]: all ruff resolved in follow on commit

File: src/axolotl/processing_strategies.py [hunk 3]
```diff
@@ -475,8 +483,237 @@ def _find_end(
...
+
+    # Move match tables to CPU lists once — per-row walk is Python-level,
+    # avoiding many small tensor accesses in the hot loop.
+    start_winner_cpu = start_winner.cpu().tolist()
+    end_match_cpu: list[Optional[list[list[bool]]]] = []
+    for em in end_match:
+        end_match_cpu.append(None if em is None else em.cpu().tolist())
+
+    # Cache boundary attributes for tight loop access.
+    bnd_role = [b.role for b in role_boundaries]
```
[the remark this answers is not in the document]
Comment (thad0ctor) [reply src/axolotl/processing_strategies.py]: removed in follow-on commit

**Comment by ved1beta** [comment]
consider adding statergies for Voxtral/Mistral3/InternVL/Glm4v too in tests , will review it shortly

**Comment by thad0ctor** [comment]
> consider adding statergies for Voxtral/Mistral3/InternVL/Glm4v too in tests , will review it shortly

@ved1beta 

I just pushed the recommended tests in 0da7285 for Voxtral, Mistral3, InternVL, and Glm4v, and added dispatch routing for the three that were missing.

**Comment by thad0ctor** [comment]
Looks like the test failures are unrelated to the PR btw

**Comment by ved1beta** [comment]
umm i can see for b < 8 , the speed up is not that effective , 
EDIT: please can you provide an end-to-end throughput measurement (tokens/sec or step-time, with/without patch, at recommended num_workers) on a real MM training run

**Comment by thad0ctor** [comment]
> umm i can see for b < 8 , the speed up is not that effective , EDIT: please can you provide an end-to-end throughput measurement (tokens/sec or step-time, with/without patch, at recommended num_workers) on a real MM training run

The scanner gains are real in my testing but it's kind of a wash if you look at it in aggreggate - I'm  on the fence - no harm if not merged.  :

**Config**: Qwen3.5-0.8B, dataset `HuggingFaceH4/llava-instruct-mix-vsft[:512]`, `chat_template: qwen3_5`, LoRA r=16, micro_batch_size=2, sequence_len=2048, `dataloader_num_workers=4`, `attn_implementation=flash_attention_2`, gradient_checkpointing, max_steps=60, seed=42, data_seed=42.

**Method**: 50 alternating trials (even *i* = patched→baseline, odd *i* = baseline→patched). Per trial: 60 measured steps via `TrainerCallback`, first 5 dropped as warmup. A/B by subprocess re-exec with `SCANNER_MODE` env-var monkey-patch before DataLoader worker fork. 9 trials failed after an external `pip install -e` swapped the editable axolotl path partway through (trials 45+); the bench skipped them. **45 trials produced fully-paired (baseline, patched) data**.

### Aggregate (pooled step-times per mode, warmup-dropped)

| metric | baseline (n=2530) | patched (n=2475) | Δ |
|---|---:|---:|---:|
| median ms/step | 262.55 | 262.01 | −0.54 ms |
| mean ms/step | 274.42 | 273.23 | −1.19 ms |
| p10 ms/step | 239.42 | 238.15 | −1.27 ms |
| p90 ms/step | 305.81 | 307.76 | +1.95 ms |

**Comment by thad0ctor** [comment]
> > umm i can see for b < 8 , the speed up is not that effective , EDIT: please can you provide an end-to-end throughput measurement (tokens/sec or step-time, with/without patch, at recommended num_workers) on a real MM training run
> 
> The scanner gains are real in my testing but it's kind of a wash if you look at it in aggreggate - I'm on the fence - no harm if not merged. :
> 
> **Config**: Qwen3.5-0.8B, dataset `HuggingFaceH4/llava-instruct-mix-vsft[:512]`, `chat_template: qwen3_5`, LoRA r=16, micro_batch_size=2, sequence_len=2048, `dataloader_num_workers=4`, `attn_implementation=flash_attention_2`, gradient_checkpointing, max_steps=60, seed=42, data_seed=42.
> 
> **Method**: 50 alternating trials (even _i_ = patched→baseline, odd _i_ = baseline→patched). Per trial: 60 measured steps via `TrainerCallback`, first 5 dropped as warmup. A/B by subprocess re-exec with `SCANNER_MODE` env-var monkey-patch before DataLoader worker fork. 9 trials failed after an external `pip install -e` swapped the editable axolotl path partway through (trials 45+); the bench skipped them. **45 trials produced fully-paired (baseline, patched) data**.
> 
> ### Aggregate (pooled step-times per mode, warmup-dropped)
> metric	baseline (n=2530)	patched (n=2475)	Δ
> median ms/step	262.55	262.01	−0.54 ms
> mean ms/step	274.42	273.23	−1.19 ms
> p10 ms/step	239.42	238.15	−1.27 ms
> p90 ms/step	305.81	307.76	+1.95 ms

More from scanner/synthesized tests:


## Environment

| | |
|---|---|
| GPU | NVIDIA GeForce RTX 5090 (32 GiB) |
| torch | 2.11.0+cu130 |
| transformers | 5.5.4 |
| flash-attn / fla-core / causal-conv1d / peft | 2.8.3 / 0.4.1 / 1.6.2.post1 / 0.19.1 |
| Model | Qwen3.5-0.8B |
| Attention | `flash_attention_2` for full-attn layers; fla + causal_conv1d fast path for linear-attn layers |
| Strategy | `Qwen3_5ProcessingStrategy` (Qwen2-VL ChatML boundaries) |
| LoRA | r=16, α=32, targets all linear projections |
| PR head | `0da7285ca` |

**A/B method.** `baseline` monkey-patches `_compute_role_keep_mask_vectorized` to fall through to the reference scanner; `patched` leaves the dispatch as the PR ships it. DataLoader workers fork after the monkey-patch is applied in main, so the change reaches them. Workers run with `torch.set_num_threads(1)` via `worker_init_fn` (matches the PR's gating predicate).

## 1. Scanner-only microbench (no DataLoader, no GPU)

100 reps + 10 warmup, `torch.set_num_threads(1)`. Representative cells.

| bs | seq_len | reference (ms) | vectorized (ms) | speedup |
|---:|--------:|---------------:|----------------:|--------:|
| 1 |   574 | 0.10 | 0.16 | 0.62× |
| 4 |  2274 | 1.41 | 1.16 | 1.21× |
| 8 |  2274 | 2.84 | 2.20 | 1.29× |
| 8 |  5674 | 7.39 | 5.31 | 1.39× |

## 2. Data-only throughput (full collator, no GPU)

10 warmup + 20 measured, median ms/batch, `num_workers=4`. Representative cells.

| bs | seq_len | baseline | patched | speedup |
|---:|--------:|---------:|--------:|--------:|
| 2 |   2048 |  3.7 |  3.3 | 1.12× |
| 4 |   4096 | 10.7 | 11.1 | 0.97× |
| 8 |   4096 | 19.8 | 19.7 | 1.01× |
| 8 |  16384 | 82.0 | 81.2 | 1.01× |

## 3. Synthetic E2E training step (median ms/step)

Custom harness, fwd+bwd+optim on Qwen3.5-0.8B + LoRA r=16, 10 warmup + 20 measured.

### seq_len ≈ 1024

| bs | nw | baseline | patched | speedup |
|---:|---:|---------:|--------:|--------:|
| 1 | 2 | 300.3 | 286.6 | 1.05× |
| 1 | 4 | 299.2 | 298.6 | 1.00× |
| 2 | 2 | 374.9 | 378.7 | 0.99× |
| 2 | 4 | 393.5 | 370.5 | 1.06× |
| 4 | 2 | 503.4 | 504.1 | 1.00× |
| 4 | 4 | 520.6 | 554.3 | 0.94× |
| 8 | 2 | 971.3 | 962.7 | 1.01× |
| 8 | 4 | 954.5 | 950.0 | 1.00× |

### seq_len ≈ 2048

| bs | nw | baseline | patched | speedup |
|---:|---:|---------:|--------:|--------:|
| 1 | 2 | 413.1 | 401.0 | 1.03× |
| 1 | 4 | 411.8 | 411.6 | 1.00× |
| 2 | 2 | 563.1 | 541.0 | 1.04× |
| 2 | 4 | 539.9 | 554.3 | 0.97× |
| 4 | 2 | 950.7 | 957.6 | 0.99× |
| 4 | 4 | 968.9 | 969.3 | 1.00× |
| 8 | 2 | OOM | OOM | — |
| 8 | 4 | OOM | OOM | — |

### seq_len ≈ 4096

| bs | nw | baseline | patched | speedup |
|---:|---:|---------:|--------:|--------:|
| 1 | 2 | 561.1 | 562.7 | 1.00× |
| 1 | 4 | 571.7 | 575.2 | 0.99× |
| 2 | 2 | 1025.8 | 973.5 | 1.05× |
| 2 | 4 | 997.5 | 1009.7 | 0.99× |
| 4+ | * | OOM | OOM | — |

### seq_len ≈ 8192

| bs | nw | baseline | patched | speedup |
|---:|---:|---------:|--------:|--------:|
| 1 | 2 | 1040.2 | 1021.0 | 1.02× |
| 1 | 4 | 1007.2 | 1004.7 | 1.00× |
| 2+ | * | OOM | OOM | — |

**Comment by winglian** [comment]
found another 2x improvement using bisect.

**Review by winglian — APPROVED** [review]

**Comment by thad0ctor** [comment]
> found another 2x improvement using bisect.

well done!
